_posts/2017-03-17-getting-rest-right.md

```diff
@@ -0,0 +1,120 @@
+---
+layout: post
+title: Getting REST right
+author: jake_hall
+---
+
+It seems everyone is building a REST API these days - this is great! It's not easy, but there are a few key things you can do to make it easy for your consumers of your API.
+
+## URLs represent a resource
+
+Try to think carefully about your URL structure before you dive in, this will save a poor REST API later on.
+
+Decide what resources you are dealing with. For these examples I'll be looking at articles, attachments and users. In my setup, articles can have attachments, but users can't, so for me it makes sense to have something similar to the following:
+
+- `GET /article` _List articles_
+- `GET /article/<id>` _Get a specific article_
+- `GET /article/<id>/attachment` _List attachments on a specific article_
+- `GET /article/<id>/attachment/<attId>` _Download a specific attachment_
+- `GET /user` _List users_
+- `GET /user/<id>` _Get a specific user_
+
+You may have noticed from the above URL schema that I have gone for singular URL nouns across the board. There is still some debate about whether singular or plural nouns are preferable, with plural seeming to win out but with most admitting it's down to individual preference.
+
+Most agree that you should stick to one or the other across the whole of your API, to keep the amount of learning to a minimum.
+
+Verbs should not show up in your API URLs - if possible. _"But what about `/search`?"_ I hear you cry.
+
+Well if search is specific to an account then maybe having something like `/article?q=search+term` may be more appropriate to limit the results of the article list. Alternatively if you do have global search endpoint, then it's probably appropriate to go for the `/search` endpoint.
+
+## Versioning
+
+Unless your API is for internal use only, and will only be used by your team, you should version your API from the first public release.
+
+There are a couple of different ways you can go about this.
+
+1. In URL versioning (http://api.example.com/v1/article/43)
+   - Probably the most popular, but unfortunately semi-breaks the "URLs represent a resource" principle
+2. Custom header (X-API-Version: v1)
+   - Developers have to learn a new header just to use your API.
+3. Accept header (Accept: application/vnd.article-v1+json)
+   - Developers can't explore your API just using a browser.
```

Comment:
Not necessarily true, most browsers will send */*  in the `Accept` header as a fallback, a good default might be to return the latest version in this case.